providers/gce/gce_instances_test.go

```diff
@@ -557,3 +558,135 @@ func TestGetZone(t *testing.T) {
 		}
 	}
 }
+
+// TestMultiProject verifies the behaviour of various functions under the
+// influence of multiProject option.
+//
+// Test setup involves creating two instances which have the same name and are
+// in the same zone but differ in their projects. A K8s Node is created with the
+// providerID pointing to the non-default project.
+//   - When multiProject is not enabled, the instance from the default project
+//     should be returned (since the project from providerID should be ignored.)
+//   - When multiProject is enabled, the instance from the non-default project
+//     should be returned (since the project from providerID would be
+//     considered.)
+func TestMultiProject(t *testing.T) {
+	defaultValues := DefaultTestClusterValues()
+	gce, err := fakeGCECloud(defaultValues)
+	require.NoError(t, err)
+
+	// instanceMap maps the instance's selfLink to the instance.
+	instanceMap := map[string]*ga.Instance{}
+
+	// Instance in the default project.
+	instanceFromDefaultProject := &ga.Instance{
+		SelfLink: fmt.Sprintf("projects/%v/zones/us-central1-c/instances/instance-1", defaultValues.ProjectID),
+		Id:       1,
+		NetworkInterfaces: []*ga.NetworkInterface{{
+			NetworkIP: "1.1.1.1",
+			StackType: "IPV4",
+			AliasIpRanges: []*ga.AliasIpRange{
+				{IpCidrRange: "10.10.10.10/24"},
+			},
+		}},
+	}
+	instanceMap[instanceFromDefaultProject.SelfLink] = instanceFromDefaultProject
+
+	// Instance in a different project.
+	nonDefaultProject := "non-default-project"
+	instanceFromNonDefaultProject := &ga.Instance{
+		SelfLink: fmt.Sprintf("projects/%v/zones/us-central1-c/instances/instance-1", nonDefaultProject),
+		Id:       2,
+		NetworkInterfaces: []*ga.NetworkInterface{{
+			NetworkIP: "2.2.2.2",
+			StackType: "IPV4",
+			AliasIpRanges: []*ga.AliasIpRange{
+				{IpCidrRange: "20.20.20.20/24"},
+			},
+		}},
+	}
+	instanceMap[instanceFromNonDefaultProject.SelfLink] = instanceFromNonDefaultProject
+	forceNonDefaultProject := cloud.ForceProjectID(nonDefaultProject)
+
+	// Setup mock response.
+	mockGCE := gce.c.(*cloud.MockGCE)
+	mi := mockGCE.Instances().(*cloud.MockInstances)
+	mi.GetHook = func(ctx context.Context, key *meta.Key, m *cloud.MockInstances, options ...cloud.Option) (bool, *ga.Instance, error) {
+		projectID := defaultValues.ProjectID
+		if len(options) == 1 && reflect.DeepEqual(options[0], forceNonDefaultProject) {
+			projectID = nonDefaultProject
+		}
+		selfLink := fmt.Sprintf("projects/%v/zones/%v/instances/%v", projectID, key.Zone, key.Name)
+		ret, ok := instanceMap[selfLink]
+		if !ok {
+			return true, nil, fmt.Errorf("instance not found")
+		}
+		return true, ret, nil
+	}
+
+	node := &v1.Node{
+		Spec: v1.NodeSpec{
+			ProviderID: fmt.Sprintf("gce://%v/us-central1-c/instance-1", nonDefaultProject),
+		},
+	}
+
+	////////////////////////////////////////////////////////////////////////////
```

Comment:
[style] We tend to not like these barber-pole comment blocks within the code statements.

Reply:
Understandable, removed.